specification.json

```diff
@@ -17,7 +17,7 @@
         {
             "id": "Requirement 1.1.2.2",
             "machine_id": "requirement_1_1_2_2",
-            "content": "The `provider mutator` function MUST invoke the `initialize` function on the newly registered provider before using it to resolve flag values.",
+            "content": "The `provider mutator` function MUST await the termination of the `initialize` function on the newly registered provider before using it to resolve flag values.",
```

Comment:
What happens if the initialize fails here?
With the current spec, I still set it and only emitted the `PROVIDER_ERROR` event.
We specified that for transient errors, the provider might "recover" and change its status to `PROVIDER_READY`.
For an error on initialize that would indicate to me that we still set the provider.

To me it is not 100% clear why we have to await the init, as we have the mechanisms to wait for and check readiness anyways. It could also be confusing, setting a provider that only eventually is set and used and until that the old provider is still resolving the flags.

Reply:
> With the current spec, I still set it and only emitted the PROVIDER_ERROR event.
We specified that for transient errors, the provider might "recover" and change its status to PROVIDER_READY.

Yes, I think this is the best implementation. The provider can recover as you say.

> To me it is not 100% clear why we have to await the init, as we have the mechanisms to wait for and check readiness anyways. It could also be confusing, setting a provider that only eventually is set and used and until that the old provider is still resolving the flags.

My goal with this is to avoid the situation where the newly set provider is used before ready when switched. Waiting for initialization means there's no period of unavailability. If a provider is switched and we don't wait for init, evaluations all start defaulting for some small period while the new provider is starting. Maybe that's something we're OK with?

Switching providers is _probably_ not something that we should consider a "first class" use case, so I'm open to simplifying things here.

Reply:
>  If a provider is switched and we don't wait for init, evaluations all start defaulting for some small period while the new provider is starting. Maybe that's something we're OK with?

Yes, my feeling is that this would be better. But no hard opinion on that iI see pros and cons of both solutions.

Reply:
Question: wait for initialization to finish: 
is there a spec for getStatus() method in SDKs or something like that ?

[spec section](https://github.com/open-feature/spec/blob/main/specification/sections/02-providers.md#requirement-242)

Reply:
> Question: wait for initialization to finish: is there a spec for getStatus() method in SDKs or something like that ?
> 
> [spec section](https://github.com/open-feature/spec/blob/main/specification/sections/02-providers.md#requirement-242)

Yes, that's basically what you've linked above. That will indicate the providers status.

Reply:
Generally speaking I think that we expect providers to be able to handle calls prior to their initialization, and the evaluation may result in some status indicating the provider wasn't ready, accompanied by a default value. Without initialization/shutdown it would have just been this way when switching providers.

I think, even with the init/shutdown additions, this is likely fine for a couple reasons.

If someone wants to switch providers, and they specifically want to use the first provider until the second provider is ready, then they can utilized named providers. Set the second provider under a different name, wait for it to be ready, and then either switch to using that name, or additionally alias that provider to the first name.

Even if a provider is not ready, you may want to be making calls to it, if those calls might be generating analytics events which are important to the operation of the specific vendor. 

I think there is potentially a nuance in provider switching that actually causes it to happen in most use cases. If the SDK has a no-op provider until a provider is set, then that is a provider switch?

Reply:
> Generally speaking I think that we expect providers to be able to handle calls prior to their initialization, and the evaluation may result in some status indicating the provider wasn't ready, accompanied by a default value. Without initialization/shutdown it would have just been this way when switching providers.
> 
> I think, even with the init/shutdown additions, this is likely fine for a couple reasons.
> 
> If someone wants to switch providers, and they specifically want to use the first provider until the second provider is ready, then they can utilized named providers. Set the second provider under a different name, wait for it to be ready, and then either switch to using that name, or additionally alias that provider to the first name.
> 
> Even if a provider is not ready, you may want to be making calls to it, if those calls might be generating analytics events which are important to the operation of the specific vendor.

I think this is very well put, and I agree. It also removes the awkward "provider was set but is not in use yet" situation. I will make changes accordingly.

@lukas-reining see above.

> I think there is potentially a nuance in provider switching that actually causes it to happen in most use cases. If the SDK has a no-op provider until a provider is set, then that is a provider switch?

No, and I _think_ that's fairly easily justified: the noop-provider isn't a _real_ provider - it's an implementation detail - a easy way of "defaulting to no-op". If the noop provider was removed from any implementation, it wouldn't violate the spec, as long as the SDK continued to noop when evaluation methods were called.

Though the noop provider exists, it's functionally equivalent to "no provider".

Reply:
> I think, the least surprise would be to return the default value. It's basically like a communication error with the feature provider. This of course comes with the risk, that features evaluated right in the beginning may always return the default value. In this case, I actually would think, that a ProviderInitialized event makes sense. This event would be emitted by the API, when the initialization is done and flag evaluation with the newly set provider can be started.

I think we generally agree not to await the initialization before setting the new provider then. With respect to the initialization event you describe, I think that the `PROVIDER_READY` event behaves in such a way. I think we have sufficient "levers and knobs" for authors with respect to initialization. I will revert the initialization point in this PR and add a bit more non-normative clarity.

Reply:
I've revered the change wrt awaiting the init event, and added a tiny bit of non-normative content as well: https://github.com/open-feature/spec/pull/193/commits/9f5a08468eb413ca42677e4a83ba4f5076ffc2e7